### PR TITLE
Organization<>Team<>User structure (closes #1540)

### LINKED ISSUE #907: Multiple, separate projects per installation
**Is your feature request related to a problem? Please describe.**
We use the events API with our mobile apps, and since there's only one API key per PostHog installation, we need to have separate installations for each app (and staging and prod versions thereof).

**Describe the solution you'd like**
I'd like to add multiple apps/projects to one PostHog deployment that get assigned separate API keys (e.g., the way Mixpanel does it work well). 

**Describe alternatives you've considered**
Only alternative is having multiple deployments of PostHog.

### LINKED ISSUE #1540: Allow creation and management of multiple teams
## Is your feature request related to a problem?

When #1468 gets merged in I want to be able to create and manage multiple entities

## Describe the solution you'd like

We probably need to move to an Organisation -> Team model, where everyone is a member of an organisation and an organisation has multiple Teams that users can switch between. For some inspiration, see [how sentry do it](https://github.com/getsentry/sentry/blob/096b355d931fad5a175d3fb06da7ed6ee0f0074a/src/sentry/models/organization.py) (note, they use 'projects' for what we call teams)

- Create Organization model (and have a specific "through" model for an OrganizationMember, as we'll want to add things like 'roles' to that relationship in the future)
- Migration that creates organisation for each Team and moves all the team members.

*If license/plan enabled, should live in ee*
- Under the switcher dropdown, show a "manage teams" option
- Allow creation of new projects

## Describe alternatives you've considered

-

## Additional context

This is a much requested features

### PR DESCRIPTION
**Pull request by Twixes**
## Changes

Taking over from #1461, this adds an Organization<>Team<>User structure, with team creation being a premium feature.

## Checklist

- [x] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [ ] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

### CODE REVIEW

**Comment by Twixes** [comment]
I believe this can be considered crucial functionality-wise done. Though a few things to get everything wrapped up nicely:
* I created an Organization > Settings page thinking that there'll be some setting there… but it's still empty, as I'm not really sure what would make sense to be moved from Team to Organization (Members, Invites, and Licenses have their own pages, so they are out of consideration). Let me know your thoughts, else the page can be removed for now (though it will surely come useful in the future).
* I absolutely **cannot figure out why org member removal refuses to work**. Looks like `member_to_delete = self.get_object()` returns something different than… the member to delete… but I'm not sure why. Please take a look, it may be completely trivial!
* TODO: At the moment only org founder can be one of its admins. I'll implement changing member levels in another PR, as this is already pretty big. Not a big deal, as nobody can remove an admin and they can't leave, so this is just similar to an owner concept for now.
* Speaking of an owner concept, I decided not to implement that in addition to "member" (level 1) and "administrator" (level 8 – the gap is if we ever want something in-between), current structure should be sufficient.
* TODO: We can now nicely replace `is_demo` Hogflix data (along with the "Delete demo data" button), by simply making Hogflix data a separate team – removable in its entirety any time.

**Comment by timgl** [comment]
Nicely done, I'll review this but already looks like some failing tests

File: posthog/migrations/0084_org_team_user.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+from django.db import migrations, models
+
+
+def forwards_func(apps, schema_editor):
+    User = apps.get_model("posthog", "User")
+    Organization = apps.get_model("posthog", "Organization")
+    OrganizationMembership = apps.get_model("posthog", "OrganizationMembership")
+    for user in User.objects.all():
+        team = user.teams_deprecated_relationship.get()
+        deterministic_derived_uuid = team.deterministic_derived_uuid
```
Comment (timgl) [inline posthog/migrations/0084_org_team_user.py:17]: This line causes running migrations locally to fail with "AttributeError: 'Team' object has no attribute 'deterministic_derived_uuid'"

**Comment by timgl** [comment]
I'm unable to test this due to the migration issue. In general, I think this pull request is too big for us to safely merge. Is there a way we could split this PR in smaller chunks without too much work?

**Comment by Twixes** [comment]
Not really, this is all very much intertwined.

**Comment by Twixes** [comment]
For posterity: these changes will be merged in a more gradual way, hence this is a draft.

**Comment by Twixes** [comment]
Some notes here:

#### Testing

- `yarn build` works for me locally, but causes Docker and E2E tests to all crash here. Webpack compilation error? The logs don't seem helpful though. 🤔 
- `test_feature_available_multi_tenancy` is the one `posthog` test that has been failing for a while and I have trouble identifying why.

Could use a fresh pair of eyes! @macobo

#### Paid functionality handling

Trying to access paid functionality in the frontend now results in an "upgrade PostHog" modal:

Screen Shot 2020-10-15 at 16 27 41

This only provides the option to email us though at the moment, enhancement ideas? @paolodamico 

#### Sidebar

Since we now have the Systems Status page, I grouped it along with Licenses to sidebar option Instance (hidden on Cloud to non-staff users).

Screen Shot 2020-10-15 at 16 29 19

**Comment by Twixes** [comment]
Code review pending resolving above two issues, but I think we can go ahead with a UX review.

**Comment by macobo** [comment]
The `yarn build` error is the following:

```
|     ERROR in ./frontend/src/scenes/sceneLogic.js
|     Module not found: Error: Can't resolve './Annotations' in '/code/frontend/src/scenes'
|      @ ./frontend/src/scenes/sceneLogic.js 120:11-122:20
|      @ ./frontend/src/scenes/App.js
|      @ ./frontend/src/index.tsx
```

Didn't go through branch history - perhaps some file got renamed and dev server never errored?

File: posthog/api/organization.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    def get_queryset(self) -> QuerySet:
+        return Organization.objects.filter(
+            id__in=OrganizationMembership.objects.filter(user=self.request.user).values("organization_id")
+        )
+
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        lookup_value = self.kwargs[self.lookup_field]
+        if lookup_value == "@current":
```
Comment (macobo) [inline posthog/api/organization.py:91]: Q: Why do we have this edge case?

The frontend should already know the relevant organization id right - this seems to result in added complexity everywhere but tests?

File: posthog/api/organization.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        read_only_fields = [
+            "id",
+            "created_at",
+            "updated_at",
+        ]
+
+    def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Organization:
+        serializers.raise_errors_on_nested_writes("create", self, validated_data)
+        request = self.context["request"]
+        with transaction.atomic():
```
Comment (macobo) [inline posthog/api/organization.py:56]: Q: How is an atomic transaction different from a normal transaction?

Comment (Twixes) [reply posthog/api/organization.py:56]: All transactions are atomic when using Postgres with Django, but `transaction.atomic()` widens the scope of atomicity from a single `INSERT/SELECT` to all queries inside the Python context, meaning if e.g. `Team.objects.create` here fails for some reason, it will be rolled back, but also `Organization.objects.create` and `OrganizationMembership.objects.create` will be rolled back.

Comment (macobo) [reply posthog/api/organization.py:56]: In postgres, all INSERT/UPDATE/DELETE queries get rolled back if the wrapping transaction fails. I guess that `.atomic` in this case is just creating a normal transaction in pg at least.

File: posthog/api/organization.py (hunk 2 above)
Comment (Twixes) [reply posthog/api/organization.py:91]: It's less work frontend side. Though I can see a way for an inconsistency to creep in if someone changes their org in another tab for instance. 🤔 Would you suggest doing away with this?

Comment (macobo) [reply posthog/api/organization.py:91]: Good point re switching organizations in tabs. If possible, I think it makes sense to remove it unless we plan to make it a API-wide pattern.

Comment (Twixes) [reply posthog/api/organization.py:91]: Definitely planning to make use of this pattern in the `/user/` endpoint: when we remake it with DRF, let's go with `/users/@me` for accessing the currently logged in user. But in the case of `/organizations` this makes less sense.

Comment (Twixes) [reply posthog/api/organization.py:91]: Decided not to go forward with that approach, as the tradeoff of occasional inconsistency here is constants delays in initial page loads, as it takes a non-negligible amount of waiting until `user` is available (and only then could additional organization- or team-specific data be loaded).

File: posthog/models/team.py [hunk 4]
```diff
@@ -106,18 +105,19 @@ class Team(models.Model):
     completed_snippet_onboarding: models.BooleanField = models.BooleanField(default=False)
     ingested_event: models.BooleanField = models.BooleanField(default=False)
     uuid: models.UUIDField = models.UUIDField(default=UUIDT, editable=False, unique=True)
+    session_recording_opt_in: models.BooleanField = models.BooleanField(default=False)
```
Comment (paolodamico) [inline posthog/models/team.py:108]: This seems completely unrelated to this PR and the UX for this is still in open discussion, I would remove for now.

File: posthog/permissions.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+from django.db.models import Model
+from rest_framework.permissions import SAFE_METHODS, BasePermission
+from rest_framework.request import Request
+
+from posthog.models import Organization, OrganizationMembership, organization
+
+CREATE_METHODS = ["POST", "PUT"]
+
+
+def extract_organization(object: Model) -> Organization:
```
Comment (paolodamico) [inline posthog/permissions.py:12]: This seems like it would go in utils maybe?

File: posthog/permissions.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    if isinstance(object, Organization):
+        return object
+    try:
+        return object.organization  # type: ignore
+    except AttributeError:
+        raise ValueError(
+            "Object not compatibile with organization-based permissions, as it does not have field `organization`!"
```
Comment (paolodamico) [inline posthog/permissions.py:19]: Minor typo

File: posthog/models/organization.py [hunk 7]
```diff
@@ -1,15 +1,33 @@
...
+        from .team import Team
+
+        with transaction.atomic():
+            organization = Organization.objects.create(**kwargs)
+            organization_membership = OrganizationMembership.objects.create(
+                organization=organization, user=user, level=OrganizationMembership.Level.ADMIN
+            )
+            team = Team.objects.create(organization=organization, **(team_fields or {}))
+            user.current_organization = organization
+            user.current_team = team
```
Comment (paolodamico) [inline posthog/models/organization.py:26]: I'm not sure about the `current_team` & `current_organization` approach of saving this stuff in the DB, seems prone to discrepancies (e.g. current team not matching current org), plus a lot of extra changes to the DB. I would consider just sending an extra header on each request indicating which team is selected on the frontend. This also makes sense if you're operating on two devices, you could easily be working with different teams on each device (e.g. I would open an incognito window if I wanted to see website numbers against app numbers).

**Review by paolodamico — CHANGES_REQUESTED** [review]
Here are my thoughts,

### General UX
- At first glance, the project/org/user selectors were somewhat confusing, won't dive too much into this because #1775 is coming, but I think a quick improvement could be naming the first project something like "Default project" (as opposed to using the org name too), that way it's a lot more clear what's going on.

image

- Accessing my settings from top bar seems to refresh the entire page instead of loading with JS.
image

- **[FYI; non-actionable]** I'll consider this new menu layout for #1775, but I definitely think we need to try to simplify this a little bit.

- The new users inviting UX is a bit convoluted, there's too many things going on. What I'd do is just allow inviting users with an email address [_rationale_: it's a common mental model for the user (simple), it's the safest of the 3, and it removes friction for the person signing up].  You would provide the invite link to the inviting user (if the instance has SMTP capabilities send an email to the invitee automatically; N.B. tell the inviting user if the email was sent). 
image

- Related to the above I would not have a separate page for invites, I would just add the invite list to the top of the member list (with a visual distinction, e.g. italics or grayed out).

- **[FYI; non-actionable]** haven't tested using the invite system pending outcome on the UX points above.


### Upgrade modal
- To make it a bit more visual (plus make it easier for the user to understand what's going on if they keep running into these modals), I would consider something like below (but @lottiecoxon might be able to help with a better graphic).
image
- For the upgrade now button I would take the user to the relevant section of the [pricing page](https://posthog.com/pricing?o=enterprise) instead of the email, we haven't provided a lot of information at this point.
- As for the copy, maybe @yakkomajuri can provide better insight, but I would remove the "Oops!" (it's not accidental/unexpected), minor adjustment ("[...] and organizations is an advanced [..]"), keep the second line in the same paragraph and change to "Upgrade now and get access to this as well as other powerful enhancements".


As for the code review just FYI I didn't repeat comments for the same stuff (e.g. using `rest_framework.status`) in multiple occurrences.


Some overall feedback. I would strongly advise against making unrelated changes in the same PR (i.e. moving the entire frontend structure tree, or migrating a ton of unrelated stuff to TS [e.g. annotations]). While they're definitely helpful at reducing tech debt, it goes against our atomic PRs workflow, bloats PR and makes it harder for reviewing (and reverting if necessary). For context, this PR introduces more line changes than the command palette or plugins PR.

File: ee/api/test/test_team.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        self.assertEqual(response_data.get("name"), "Default")
+        self.assertEqual(self.organization.teams.count(), 2)
+
+    def test_delete_team_own_second(self):
+        team = Team.objects.create(organization=self.organization)
+        response = self.client.delete(f"/api/projects/{team.id}")
+        self.assertEqual(response.status_code, 204)
+        self.assertEqual(Team.objects.filter(organization=self.organization).count(), 1)
+
+    def test_no_delete_last_team(self):
```
[the remark this answers is not in the document]
Comment (Twixes) [reply ee/api/test/test_team.py:27]: Discussed this with @timgl. Having last org/project be deletable would invalidate a lot of our current assumptions, because it's always been assumed that `User.team` is never `None`. For example in such cases most sidebar items would have to be hidden, endpoints would have to be disabled. I think we should make this flexible in the future, although I'd preface that with #1775.

File: ee/api/test/test_organization.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        organization, organization_membership, team = Organization.objects.bootstrap(self.user)
+        organization_membership.level = OrganizationMembership.Level.MEMBER
+        organization_membership.save()
+        self.assertTrue(Organization.objects.filter(id=organization.id).exists())
+        self.assertTrue(Team.objects.filter(id=team.id).exists())
+        response = self.client.delete(f"/api/organizations/{organization.id}")
+        self.assertEqual(response.status_code, 403)
+        self.assertTrue(Organization.objects.filter(id=organization.id).exists())
+        self.assertTrue(Team.objects.filter(id=team.id).exists())
+
```
[the remark this answers is not in the document]
Comment (Twixes) [reply ee/api/test/test_organization.py:50]: That's covered by `test_organization.py` in the `posthog` (FOSS-only) app.

File: posthog/api/organization.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    viewsets,
+)
+from rest_framework.generics import DestroyAPIView
+from rest_framework.mixins import DestroyModelMixin
+
+from posthog.models import Organization, OrganizationMembership, Team, team
+from posthog.permissions import CREATE_METHODS, OrganizationAdminWritePermissions, OrganizationMemberPermissions
+
+
+class PremiumMultiorganizationPermissions(permissions.BasePermission):
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization.py:23]: It seems like an actual permissions thing to me (especially since 402 is very rarely used) plus permission classes are very nice to use. Not sure if this is doable in a similarly simple way without permission classes.

File: posthog/api/organization.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+                raise exceptions.ValidationError(
+                    f"Cannot remove organization since that would leave member {member.email} organization-less!"
+                )
+        self.perform_destroy(instance)
+        return response.Response(status=status.HTTP_204_NO_CONTENT)
+
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        lookup_value = self.kwargs[self.lookup_field]
+        if lookup_value == "@current":
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization.py:88]: Basically my concern is that we'd have to run `check_object_permissions` twice then.

File: posthog/api/organization_member.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+from posthog.permissions import OrganizationAdminWritePermissions, OrganizationMemberPermissions, extract_organization
+
+
+class OrganizationAdminWriteExcludingDeletePermissions(BasePermission):
+    """Require organization admin level to change object, allowing everyone read AND delete."""
+
+    message = "Your organization access level is insufficient."
+
+    def has_object_permission(self, request: Request, view, object: Model) -> bool:
+        if request.method in SAFE_METHODS or request.method == "DELETE":
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:19]: What do you mean by stuff?

File: posthog/api/organization.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            return self.request.user.organization
+        filter_kwargs = {self.lookup_field: lookup_value}
+        obj = get_object_or_404(queryset, **filter_kwargs)
+        self.check_object_permissions(self.request, obj)
+        return obj
+
+    def get_queryset(self) -> QuerySet:
+        return Organization.objects.filter(
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization.py:96]: I think `self.request.user.organizations.all()`, but :+1: probably a good idea.

File: posthog/api/organization_member.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+            return True
+        organization = extract_organization(object)
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:21]: Basically just trusting the PostHog developer putting in this check to know whether their object is organization-related.

File: posthog/api/organization_member.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        return (
+            OrganizationMembership.objects.get(user=request.user, organization=organization).level
+            >= OrganizationMembership.Level.ADMIN
+        )
+
+
+class OrganizationMembershipSerializer(serializers.ModelSerializer):
+    user_first_name = serializers.CharField(source="user.first_name", read_only=True)
+    user_email = serializers.CharField(source="user.email", read_only=True)
+    membership_id = serializers.CharField(source="id", read_only=True)
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:31]: Because the class is actually misnamed, should be `OrganizationMemberSerializer`. It's a hybrid between `User` and `OrganizationMembership`.

File: posthog/api/organization.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (paolodamico) [reply posthog/api/organization.py:96]: Right, yes!

File: posthog/api/organization_member.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+    class Meta:
+        model = OrganizationMembership
+        fields = ["membership_id", "user_id", "user_first_name", "user_email", "level", "joined_at", "updated_at"]
+
+
+class OrganizationMemberViewSet(
+    NestedViewSetMixin,
+    mixins.DestroyModelMixin,
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:40]: More verbose about intentions this way.

File: posthog/api/organization_member.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (paolodamico) [reply posthog/api/organization_member.py:19]: So the permission validates that anyone can do any safe method (GET, HEAD, OPTIONS) but DELETE too, why allow DELETE is my question

File: posthog/api/organization_member.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    mixins.UpdateModelMixin,
+    mixins.ListModelMixin,
+    viewsets.GenericViewSet,
+):
+    serializer_class = OrganizationMembershipSerializer
+    pagination_class = None
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:46]: Good point. Well, handling pagination is a bit more work on the frontend, but important to keep response structure consistent.

File: posthog/api/organization_member.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:19]: Ah, yup, it used to be that `destroy` in the viewset itself checked if the user is only trying to delete themselves (which is allowed), not someone else (which is disallowed) – but good point, move that to this permission class so that it's purpose is clearer.

File: posthog/models/team.py (hunk 4 above)
Comment (Twixes) [reply posthog/models/team.py:108]: Ah, this is actually not new at all, I just moved this line to be above the deprecated fields (as there are new such ones).

File: posthog/api/test/base.py [hunk 18]
```diff
@@ -25,38 +25,38 @@ def setUp(self):
...
     pass
 
 
-class TransactionBaseTest(TestMixin, TransactionTestCase, ErrorResponsesMixin):
+class TransactionBaseTest(TestMixin, ErrorResponsesMixin, TransactionTestCase):
     pass
 
 
-class APIBaseTest(APITestCase, ErrorResponsesMixin):
+class APIBaseTest(ErrorResponsesMixin, APITestCase):
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/test/base.py:59]: Just consistency with the ordering in classes above.

File: posthog/api/teams.py [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+        OrganizationAdminWritePermissions,
+        PremiumMultiprojectPermissions,
+    ]
+    lookup_field = "id"
+    ordering_fields = ["created_by"]
+    ordering = ["-created_by"]
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        if isinstance(self.request.user, AnonymousUser):
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/teams.py:104]: Yup, obviously. 👍

File: posthog/api/organization_member.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+    def filter_queryset_by_parents_lookups(self, queryset) -> QuerySet:
+        parents_query_dict = self.get_parents_query_dict()
+        if parents_query_dict:
+            if parents_query_dict["organization_id"] == "@current":
+                parents_query_dict["organization_id"] = self.request.user.organization.id
+            try:
+                return queryset.filter(**parents_query_dict)
+            except ValueError:
+                raise Http404
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:61]: Good point, this is a holdover from copying a large part of the method from the `NestedViewSetMixin` superclass (drf-extensions).

File: posthog/api/organization_member.py [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    ordering = ["level", "-joined_at"]
+
+    def filter_queryset_by_parents_lookups(self, queryset) -> QuerySet:
+        parents_query_dict = self.get_parents_query_dict()
+        if parents_query_dict:
+            if parents_query_dict["organization_id"] == "@current":
+                parents_query_dict["organization_id"] = self.request.user.organization.id
+            try:
+                return queryset.filter(**parents_query_dict)
+            except ValueError:
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:60]: Hm, this is how drf-exceptions handles it and it seems pretty trustworthy.  🤔

File: posthog/api/teams.py [hunk 22]
```diff
@@ -0,0 +1,178 @@
+import json
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/teams.py:1]: On one hand, plural is more RESTful and I'd prefer we move to plural overall, on the other hand there'd be consistency in between. So you suggest sticking to singular?

File: posthog/permissions.py (hunk 5 above)
Comment (Twixes) [reply posthog/permissions.py:12]: It's very very permissions specific though, I feel like it'd just bloat utils when permissions is where it makes the most sense.

File: posthog/models/organization.py (hunk 7 above)
Comment (Twixes) [reply posthog/models/organization.py:26]: I agree that that would be great, but it would also be a significantly larger refactor (consider insights, which would need to be project-nested then like so: `/api/projects/<project_id>/insights/` – this would have to apply to everything). Could we do this a bit later in another PR? One of the points here is that this one is already too big (which is also true).

File: posthog/api/organization_member.py [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+        obj = get_object_or_404(queryset, **filter_kwargs)
+        self.check_object_permissions(self.request, obj)
+        return obj
+
+    def destroy(self, request, *args, **kwargs):
+        """Member removal with validation (admin permissions)."""
+        membership_to_delete = self.get_object()
+        try:
+            if (
+                membership_to_delete.user != request.user
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:80]: Moved this to the permission class. Unsure how to standardize.

File: ee/api/test/test_organization.py [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+    def test_no_delete_organization_not_belonging_to(self):
+        organization = Organization.objects.create(name="Some Other Org")
+        response = self.client.delete(f"/api/organizations/{organization.id}")
+        self.assertEqual(response.status_code, 403)
```
[the remark this answers is not in the document]
Comment (Twixes) [reply ee/api/test/test_organization.py:54]: That would also make sense, although that'd require throwing the exception in a less elegant and reusable way than the permission class.

File: posthog/api/__init__.py [hunk 25]
```diff
@@ -16,16 +13,21 @@
...
     paths,
     person,
     personal_api_key,
-    team_user,
+    teams,
 )
 
 
-class OptionalTrailingSlashRouter(routers.DefaultRouter):
+class HedgeRouter(ExtendedDefaultRouter):
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/__init__.py:26]: True, made it less tongue-in-cheek.

File: posthog/api/organization_member.py [hunk 26]
```diff
@@ -0,0 +1,88 @@
+from django.db.models import Model, QuerySet, query
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/organization_member.py:1]: Hmm, interesting point, could you elaborate why unnecessary?

File: ee/api/test/test_organization.py [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+
+from posthog.models.organization import Organization, OrganizationInvite, OrganizationMembership
+from posthog.models.team import Team
+
+from .base import APILicensedTest
+
+
+class TestOrganizationEnterpriseAPI(APILicensedTest):
+    def test_create_organization(self):
+        response = self.client.post("/api/organizations/", {"name": "Test"})
```
[the remark this answers is not in the document]
Comment (Twixes) [reply ee/api/test/test_organization.py:14]: Any signed in user. Now, this is actually pretty clear on self-hosted, where the license is per-instance and we can just block creating additional organizations without a license, but unsure what to do on Cloud, where billing is per-organization. So we can't really make multiple organizations a paid feature. And also can't make multiple projects then, because you can create an organization as a surrogate of that anyway. Sentry for instance, which is quite similar in how its built, puts no restrictions on org/proj creation. Thoughts?

File: frontend/src/scenes/UpgradeModal.tsx [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+export function UpgradeModal(): JSX.Element {
+    const { upgradeModalFeatureName } = useValues(sceneLogic)
+    const { hideUpgradeModal } = useActions(sceneLogic)
+
+    return (
+        <Modal
+            title="Unleash PostHog's Full Power"
+            okText="Upgrade Now"
+            cancelText="Maybe Later"
+            onOk={() => {
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/UpgradeModal.tsx]: Fortunately since we're using Kea anyway, I added listeners with `posthog.capture`, so this is covered.

File: ee/api/test/test_organization.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (paolodamico) [reply ee/api/test/test_organization.py:14]: Well regarding on whether the feature should be restricted, I think at the very least we should start enabling it from the Starter plan, might even make sense to push to Hobby. @timgl thoughts?

As for creating orgs, what you described makes sense.

File: posthog/models/organization.py (hunk 7 above)
Comment (paolodamico) [reply posthog/models/organization.py:35]: I was considering something easier, just add a header (e.g. `Posthog-Project`), on the backend it's a simple mixin to parse that and on the frontend I believe you can use the `api` lib that all requests use (or an interceptor). I do think this might simplify the current PR, but obviously not dead set against doing it this way, we can definitely consider this for an improvement later on.

File: ee/api/test/test_team.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (paolodamico) [reply ee/api/test/test_team.py:27]: Sounds good 👌

**Review by paolodamico — CHANGES_REQUESTED** [review]
Some general comments (mostly UX), obviously please feel free to push back in any of them, or disregard them altogether if you think it's best:

### Org Invites
- I still think it's still a lot better to have the invites in the same page as members, it's confusing to have that separate and adds an extra unnecessary view. It also reminds the user about stale invites so they can take action (either cancel it or ping their invitee).
- I would change the "Create an invite" button to "Invite team member" or "Invite user" it's more straightforward this way.
- The ID scheme is somewhat confusing, it took me a little while to understand what was going on. The ID is also a link to do the signup but it doesn't make sense because if the inviting user clicks it, it redirects them to the insights page (on account of being logged in). My suggestion to improve this: Do not show the ID altogether, it's meaningless for the user. On the modal, after submitting the form, provide the full invite link with instructions and a copy to clipboard button. _FYI I'm creating an issue so that we can add automatic emails as a later improvement_.
- With the new invite member mental model, the "Uses so Far" and "Last Used by" don't make sense anymore. The behavior you would expect is just to see the unused invites in the list. After the invitee has signed up, the invite would automatically be deleted.
- **[Functionality]** - I was able to use a sign up link with a different email (see below).
screenshot

- I don't think I should be able to delete myself (if I remember correctly this was restricted at the API level).
image

- **[Functionality]** - Attempting to remove a team member made React crash (see below), but the member was indeed removed.
    - Related to the above, I was already logged in with the deleted member, I refreshed the page and got a 500 error ("Organization matching query does not exist."). I would strongly suggest just permanently deleting the user and setting foreign keys to `None` to indicate that the related objects were created/updated by a deleted user.
image

### Creating projects & orgs
- After I created a project I was taken to the ingestion guided setup. It's a bit weird because it feels like I'm going through the initial setup again. I would suggest taking the user to the project page or something like that, and give them the option to enter the guided ingestion process independently. Definitely not force them on them because now I can't do anything (i.e. go back to my older project) until I finish this which I may not be able to do right now.
- Went through the ingestion process. The instant I got to the last step it showed me a message that I had already ingested events (without doing anything), plus after clicking "Complete", it took me back to the beginning of the ingestion process, I couldn't leave this loop at all. Same problem when creating an org.
image

- Switching between projects is a bit unreliable (project is not always changed). Same for switching between orgs (see below).

image

File: frontend/src/layout/TopContent/CommandPaletteButton.tsx [hunk 29]
```diff
@@ -16,7 +16,7 @@ export function CommandPaletteButton(): JSX.Element {
             title={isPaletteShown ? 'Hide Command Palette' : 'Show Command Palette'}
         >
             <SearchOutlined size={1} style={{ marginRight: '0.5rem' }} />
-            {platformCommandControlKey('K')}
+            <b>{platformCommandControlKey('K')}</b>
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/layout/TopContent/CommandPaletteButton.tsx:19]: Agreed that bold 700 isn't perfect here, but regular 400 on the other hand is hardly noticeable. Made it medium 500.

File: frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts [hunk 30]
```diff
@@ -495,9 +489,39 @@ export const commandPaletteLogic = kea<
...
+                            push('/project/settings')
+                        },
+                    },
+                    {
+                        icon: SmileOutlined,
+                        display: 'Go to My Settings',
+                        synonyms: ['account'],
                         executor: () => {
-                            push('/team')
+                            push('/me/settings')
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/lib/components/CommandPalette/commandPaletteLogic.ts:524]: "Me" seems more distinctly personal to me than "Account". "Profile" on the other hand sounds like something public.

File: frontend/src/scenes/organization/Invites/logic.tsx [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+            createInvite: async ({ targetEmail }: { targetEmail?: string }) => {
+                const newInvite = await api.create('api/organizations/@current/invites/', { target_email: targetEmail })
+                return [newInvite, ...values.invites]
+            },
+            deleteInvite: async (invite) => {
+                await api.delete(`api/organizations/@current/invites/${invite.id}/`)
+                toast(
+                    <div>
+                        <h1 className="text-success">
+                            <CheckCircleOutlined /> Invite <b>{invite.id}</b> successfully removed!
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/organization/Invites/logic.tsx:25]: Replaced with email.

File: frontend/src/scenes/project/Settings/index.js [hunk 32]
```diff
@@ -29,46 +23,40 @@ function _Setup() {
...
             variable instead of hard-coding.
             <Divider />
-            <h2 id="personal-api-keys">Personal API Keys</h2>
-            <PersonalAPIKeys />
-            <Divider />
             <h2 id="urls">Permitted Domains/URLs</h2>
-            These are the domains and URLs where the Toolbar will automatically open if you're logged in. It's also
-            where you'll be able to create Actions.
+            <p>
+                These are the domains and URLs where the Toolbar will automatically open if you're logged in. It's also
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/project/Settings/index.js:52]: Not sure how to reword it atm, maybe add this to #1755.

File: frontend/src/scenes/sceneLogic.js [hunk 33]
```diff
@@ -45,32 +49,38 @@ export const routes = {
     '/actions/live': 'liveActions',
     '/actions': 'actions',
     '/insights': 'insights',
-    '/setup': 'setup',
     '/events': 'events',
+    '/sessions': 'sessions',
     '/person_by_id/:id': 'person',
     '/person/*': 'person',
-    '/people': 'people',
+    '/people/persons': 'people',
```
[the remark this answers is not in the document]
Comment (Twixes) [reply frontend/src/scenes/sceneLogic.js:56]: Keeping URLs consistent with the sidebar - and Persons is part of People.

**Comment by Twixes** [comment]
Good points @paolodamico, to expand on a couple:

- I'll push back on merging Members and Invites with the current settings layout, because I feel like that wouldn't work great stylistically. Keen on doing this when we reorganize the sidebar though, great point on stale invites.
  On a related note, made invites delete themselves on use (since they're only per-email now).
- Deleting yourself is basically just leaving. Though maybe administrators should not be allowed to leave, just so there never is an admin-less org. 🤔 The owner concept comes in useful here, maybe should introduce that.

Resolved all the other points.
